src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/ProjectDiagnosticSource.cs

```diff
@@ -11,28 +11,59 @@
             : null;
+    public string ToDisplayString() => Project.Name;
 
-    public async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(
-        IDiagnosticAnalyzerService diagnosticAnalyzerService,
-        RequestContext context,
-        CancellationToken cancellationToken)
+    private sealed class FullSolutionAnalysisDiagnosticSource(Project project, Func<DiagnosticAnalyzer, bool>? shouldIncludeAnalyzer) : ProjectDiagnosticSource(project)
     {
-        // Directly use the IDiagnosticAnalyzerService.  This will use the actual snapshots
-        // we're passing in.  If information is already cached for that snapshot, it will be returned.  Otherwise,
-        // it will be computed on demand.  Because it is always accurate as per this snapshot, all spans are correct
-        // and do not need to be adjusted.
-        var projectDiagnostics = await diagnosticAnalyzerService.GetProjectDiagnosticsForIdsAsync(Project.Solution, Project.Id,
-            diagnosticIds: null, ShouldIncludeAnalyzer, includeSuppressedDiagnostics: false, includeNonLocalDocumentDiagnostics: true, cancellationToken).ConfigureAwait(false);
-        return projectDiagnostics;
+        private readonly Func<DiagnosticAnalyzer, bool>? _shouldIncludeAnalyzer = shouldIncludeAnalyzer;
+
+        public override async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(
+            IDiagnosticAnalyzerService diagnosticAnalyzerService,
+            RequestContext context,
+            CancellationToken cancellationToken)
+        {
+            // Directly use the IDiagnosticAnalyzerService.  This will use the actual snapshots
+            // we're passing in.  If information is already cached for that snapshot, it will be returned.  Otherwise,
+            // it will be computed on demand.  Because it is always accurate as per this snapshot, all spans are correct
+            // and do not need to be adjusted.
+            return await diagnosticAnalyzerService.GetProjectDiagnosticsForIdsAsync(Project.Solution, Project.Id,
+                diagnosticIds: null, _shouldIncludeAnalyzer, includeSuppressedDiagnostics: false, includeNonLocalDocumentDiagnostics: false, cancellationToken).ConfigureAwait(false);
+        }
     }
 
-    public string ToDisplayString() => Project.Name;
+    private sealed class CodeAnalysisDiagnosticSource(Project project, ICodeAnalysisDiagnosticAnalyzerService codeAnalysisService) : ProjectDiagnosticSource(project)
+    {
+        private readonly ICodeAnalysisDiagnosticAnalyzerService _codeAnalysisService = codeAnalysisService;
+
+        public override Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(
+            IDiagnosticAnalyzerService diagnosticAnalyzerService,
+            RequestContext context,
+            CancellationToken cancellationToken)
+        {
+            return _codeAnalysisService.GetProjectDiagnosticsAsync(Project.Id, Project.Solution.Workspace, cancellationToken);
```

Comment:
nit: this API could just take a Project, instead of Workspace+Id.  But i dont' feel strongly.  If it takes the latter, i'd prefer it say "GetLastComputedProjectDiagnosticsAsync".  That way it is clear this is an API that is not computing up to date results.